### PR TITLE
Multiple minor Firestore changes

### PR DESCRIPTION
**Pull request by jskeet**
Suggest reviewing one commit at a time

### CODE REVIEW

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/QuerySnapshotTest.cs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+            Assert.Equal(2, querySnapshot.Count);
+            // Indexer
+            Assert.Same(doc1, querySnapshot[0]);
+            Assert.Same(doc2, querySnapshot[1]);
+            Assert.Throws<ArgumentOutOfRangeException>(() => querySnapshot[-1]);
+            Assert.Throws<ArgumentOutOfRangeException>(() => querySnapshot[2]);
+            // Use IEnumerable<DocumentSnapshot>
+            Assert.Equal(docs, querySnapshot);
+            // Use IEnumerable
+            Assert.Equal(docs, querySnapshot.Cast<object>().Cast<DocumentSnapshot>());
```
Comment (chrisdunelm) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/QuerySnapshotTest.cs]: nit: why the double cast? Why not a single `Cast()` to `DocumentSnapshot`?

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/Utilities/EqualityHelpers.cs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        /// <summary>
+        /// Computes a hash code for a list
+        /// </summary>
+        internal static int GetListHashCode<T>(IReadOnlyList<T> list) where T : IEquatable<T>
+        {            
+            if (list == null)
+            {
+                return 0;
+            }
+            unchecked
```
Comment (chrisdunelm) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore/Utilities/EqualityHelpers.cs]: nit: unchecked is the default, so is there a need to explicitly use it?

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/QuerySnapshotTest.cs (hunk 1 above)
Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/QuerySnapshotTest.cs]: It's an attempt to confuse any optimization LINQ might have that notices it's already the right iterable type. I'm not 100% sure it's effective, but I'm hopeful.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/Utilities/EqualityHelpers.cs (hunk 2 above)
Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore/Utilities/EqualityHelpers.cs]: While it's the default, I suspect we may well be tempted to make checked the default at some point... but hash combinations should *never* be done in a checked context.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/QuerySnapshotTest.cs (hunk 1 above)
Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/QuerySnapshotTest.cs]: I think OfType will work better for this, actually:
https://codeblog.jonskeet.uk/2011/01/13/reimplementing-linq-to-objects-part-33-cast-and-oftype/

**Pull request merged**